### DISCUSSION TITLE
How do I hide MacOS Title Bar

### DISCUSSION THREAD

**Discussion by duvenagep** [slint-ui/slint#4284] [Q&A] [answered]
Hi 

I am very new to Slint but in an attempt to understand the concepts better, I thought I would build the MacOs calculator app.

I got far very quickly (kudos 🙏 ) however, I got stuck on something small, and after hours with the documentation I thought perhaps I was missing something.

Ignoring some font settings I was able to make an *almost* exact clone:

| MacOS | Slint-Clone |
|-------------|--------------|
| <img width="243" alt="Screenshot 2023-12-31 at 18 50 50" src="https://github.com/slint-ui/slint/assets/13404204/5fe0f3ed-d923-4c07-8135-19d1294125bf">     | <img width="242" alt="Screenshot 2023-12-31 at 18 52 52" src="https://github.com/slint-ui/slint/assets/13404204/829b9aa6-e780-4634-8300-a76cb1f6894f">      |

What I am trying to do is hide the title bar (or more specifically make my content a full-screen view). 
How can this be achieved?

Also is it possible to give the entire window and opacity similar to what can be given to a component?

I had previously attempted this MacOS Calculator clone using the Iced library and I was able to get the desired effect (see below) so I assume this must be easily achievable using Slint. Any help would be greatly appreciated.
| MacOS | Iced-Clone |
|-------------|--------------|
| <img width="243" alt="Screenshot 2023-12-31 at 18 50 50" src="https://github.com/duvenagep/iced-calculator/assets/13404204/66ac1f25-1aa7-4b57-98de-03baffb06ad9">     | <img width="242" alt="Screenshot 2023-12-31 at 18 52 52" src="https://github.com/duvenagep/iced-calculator/assets/13404204/7d4bc364-6f49-4afe-bcb9-3a2ef2d34363">      |

**Comment by tronical** [comment]
Hey, that looks good and very close :). We don't have a stable API yet to hide the title bar, but what you can do alternatively is to use the private API by means of using the [i-slint-backend-winit](https://docs.rs/i-slint-backend-winit/latest/i_slint_backend_winit/) crate. That gives you access to the `winit::window::Window`, which in turn gives you access to the `NSWindow`, and then I think you can toggle the [`titleVisibility`](https://developer.apple.com/documentation/appkit/nswindow/1419635-titlevisibility?language=objc). It requires a bit of platform specific low-level code, but it should be possible.

**Reply by duvenagep** [reply]
Hi @tronical 

Thanks for pointing me in the right direction! I will investigate further. 👍

**Reply by bjorn** [reply]
Normally `winit` provides an abstraction, though only available on macOS, in the form of [`WindowBuilderExtMacOS::with_titlebar_transparent`](https://github.com/rust-windowing/winit/blob/816798bfd152194fc35b2de263887b3dca6205f3/src/platform/macos.rs#L220) (somehow missing from [the docs](https://docs.rs/winit/0.29.9/winit/platform/index.html)). However, I don't think Slint currently makes it possible (or easy) to hook into the window creation? I didn't see a way to change the title bar visibility after window creation through winit API, so maybe accessing the `NSWindow` is needed for that.

**Reply by tronical** [reply]
Olivier just implemented a feature in the `i-slint-backend-winit` crate that gives access to the `WindowBuilder` (and by extension `WindowBuilderExtMacOS`. If you use the master branch of Slint from git, you can give it a try. You can find an example snippet in the snapshot docs: https://slint.dev/snapshots/master/docs/rust/i_slint_backend_winit/struct.Backend#structfield.window_builder_hook

**Reply by ogoffart** [reply]
(the change was https://github.com/slint-ui/slint/pull/4291)

> somehow missing from the docs

It is in the docs, but you need to select the platform to be apple-darwin:  https://docs.rs/winit/latest/x86_64-apple-darwin/winit/platform/macos/trait.WindowBuilderExtMacOS.html#tymethod.with_title_hidden

**Reply by duvenagep** [reply]
Thanks @tronical @ogoffart 

I just updated my project and will do some experimentation tonight! If I get it working I will post the result here.

Btw, thanks for the quick response, really makes using this library amazing! 🙏

**Reply by duvenagep** [reply]
Hi @tronical @ogoffart 

In an attempt to test the new functionality I added slint via the git. However I now get this error and I am struggling to to figure out how to set the renderer.
<img width="1508" alt="Screenshot 2024-01-09 at 23 49 28" src="https://github.com/slint-ui/slint/assets/13404204/362c718c-3481-42b6-8701-3a20b582fe55">

I tried setting the SLINT_BACKEND env variable as per the docs but still the same error persists. 

Any advice on how to resolve this issue?

**Reply by ogoffart** [reply]
You need to add one of the `renderer-...` features in the Cargo.toml

**Reply by duvenagep** [reply]
Hi @ogoffart 

Perhaps I am missing something really obvious but when I add Slint directly from github via `https://github.com/slint-ui/slint.git` there are no features available for rendering other than below:

![Screenshot 2024-01-10 at 09 08 57](https://github.com/slint-ui/slint/assets/13404204/6aefa37e-7e28-4abe-bc6f-c888ff347482)

These only become available via `cargo add slint --features renderer-...`

Below are my current dependencies in .TOML, could you please point me in the right direction?

![Screenshot 2024-01-10 at 09 20 33](https://github.com/slint-ui/slint/assets/13404204/42ac9624-a0f7-4938-8777-d0b7150e6f2c)

**Reply by ogoffart** [reply]
you should use the same version for sint and )-slint-backend-winit:  both  `{ git = "https://github.com/slint-ui/slint" }`

**Reply by duvenagep** [reply]
Got it thanks!

**Reply by duvenagep** [reply, marked as answer]
@tronical , @ogoffart 

Thanks for all the help. Below the final results since the API change.

## Results

| MacOS | Slint-Clone |
|-------------|--------------|
| <img width="243" alt="Screenshot 2023-12-31 at 18 50 50" src="https://github.com/slint-ui/slint/assets/13404204/56f5a4ce-731b-4a78-a0e3-eb089662fbbe">     | <img width="242" alt="Screenshot 2023-12-31 at 18 52 52" src="https://github.com/slint-ui/slint/assets/13404204/0d35905e-131f-4174-83e5-f1a8b20b8b5b">      |


Also thought I would put the code snippet here for others if needed.
```rust
use i_slint_backend_winit::winit::platform::macos::WindowBuilderExtMacOS;

slint::include_modules!();

fn main() -> Result<(), slint::PlatformError> {
    let mut backend = i_slint_backend_winit::Backend::new().unwrap();
    backend.window_builder_hook = Some(Box::new(|builder| {
        builder
            .with_fullsize_content_view(true)
            .with_title_hidden(true)
            .with_titlebar_transparent(true)
    }));

    slint::platform::set_platform(Box::new(backend));
    let ui = AppWindow::new()?;
    let calc = ui.window();
    ui.run()
}

```

**Reply by aurexav** [reply]
Do you know how to get the title bar size? I want to set a padding base on that value.

**Reply by ogoffart** [reply]
Maybe you can get is using bindings to the `cocoa` crate. But I'm not familiar with that.
